README
NAME
    Test::Inline - Inlining your tests next to the code being tested

DESCRIPTION
    Embedding tests allows tests to be placed near the code it's testing.

    This is a nice supplement to the traditional .t files.

    It's like XUnit, only better and Perl-style.

  How does it work?
    "Test::Inline" lets you write small fragments of general or
    method-specific testing code, and insert it anywhere you want in your
    modules, inside a specific tagged POD segment, like the following.

      # A fragment of general test code
  
      =begin testing
  
      ok( -f /proc/cpuinfo, 'Host has a standard /proc/cpuinfo file' );
  
      =end testing
  
      # Completely test a single method
  
      =begin testing label
  
      # Test generation of the <label> HTML tag
      is( My::HTML->label('foo'),        '<label>foo</label>',           '->label(simple) works' );
      is( My::HTML->label('bar', 'foo'), '<label for="bar">foo</label>', '->label(for) works'    );
  
      =end testing

    You can add as many, or as few, of these chunks of tests as you wish.
    The key condition when writing them is that they should be conceptually
    independant of each other. Each chunk of testing code should not die or
    crash if it is run before or after another chunk.

    Using inline2test or another test compiler, you can then transform these
    chunks in a test script, or an entire tree of modules into a complete
    set of standard Test::More-based test scripts.

    These test scripts can then be executed as normal.

  What is Test::Inline good for?
    "Test::Inline" is incredibly useful for doing ad-hoc unit testing.

    In any large groups of modules, you can add testing code here, there and
    everywhere, anywhere you want. The next time the test compiler is run, a
    new test script will just appear.

    It's also extremely for systematically testing self-contained code.

    That is, any code which can be independantly tested without the need for
    external systems such as databases, and that has no side-effects on
    external systems.

    All of this code, written by multiple people, can then have one single
    set of test files generated. You can check all the bits and pieces of a
    large API, or anything you like, in fine detail.

  What is Test::Inline bad for?
    "Test::Inline" is not a complete testing solution, and there are several
    types of testing you probably DON'T want to use it for.

    *   Static testing across the entire codebase

    *   Functional testing

    *   Tests with side-effects such as those that might change a testing
        database

  Getting Started
    Because Test::Inline creates test scripts that _don't_ start with a
    numeric order, is to create your normal tests using file names in the
    CPAN style of "01_compile.t", "02_main.t", "03_foobar.t".

    You can then add your testing fragments wherever you like throughout
    your code, and then use the "inline2test" script to generate the test
    scripts for the inline tests. By default the test will be named after
    the packages that the test fragments are in.

    Tests for Class::Name will end up in the file "class_name.t".

    These test files sit quite happily alongside your number test scripts.

    When you run the test suite as you normally would, the inline scripts
    will be run after the numbered tests.

METHODS
METHODS
  new
      my $Tests = Test::Inline->new(
              verbose  => 1,
              output   => 'auto',
              manifest => 'auto.manifest',
              );

    The "new" constructor creates a new test generation framework. Once the
    constructor has been used to create the generator, the "add_class"
    method can be used to specify classes, or class heirachies, to generate
    tests for.

    verbose - The "verbose" option causes the generator to write state and
    debugging information to STDOUT as it runs.

    manifest - The "manifest" option, if provided, will cause a manifest
    file to be created and written to disk. The manifest file contains a
    list of all the test files generated, but listed in the prefered order
    they should be processed to best satisfy the class-level dependency of
    the tests.

    check_count - The "check_count" value controls how strictly the test
    script will watch the number of tests that have been executed.

    When set to false, the script does no count checking other than the
    standard total count for scripts (where all section counts are known)

    When set to 1 (the default), "Test::Inline" does smart count checking,
    doing section-by-section checking for known-count sections only when the
    total for the entire script is not known.

    When set to 2 or higher, "Test::Inline" does full count checking, doing
    section-by-section checking for every section with a known number of
    tests.

    file_content - The "file_content" option should be provided as a CODE
    reference, which will be passed as arguments the "Test::Inline" object,
    and a single Test::Inline::Script object, and should return a string
    containing the contents of the resulting test file. This will be written
    to the "OutputHandler".

    output - The "output" option provides the location of the directory
    where the tests will be written to. It should both already exist, and be
    writable. If using a custom "OutputHandler", the value of "output"
    should refer to the location within the OutputHandler the files will be
    written to.

    InputHandler - The "InputHandler" option, if provided, supplies an
    alternative "FileHandler" from which source modules are retrieved.

    OuputHandler - The "OutputHandler" option, if provided, supplies an
    alternative "FileHandler" to which the resulting test scripts are
    written.

    Returns a new "Test::Inline" object on success.

    Returns "undef" if there is a problem with one of the options.

  InputHandler
    The "InputHandler" method returns the file handler object that will be
    used to find and load the source code.

  ExtractHandler
    The "ExtractHandler" accessor returns the object that will be used to
    extract the test sections from the source code.

  ContentHandler
    The "ContentHandler" accessor return the script content generation
    handler.

  OutputHandler
    The "OutputHandler" accessor returns the file handler object that the
    generated test scripts will be written to.

  add $file, $directory, \$source, $Handle
    The "add" method is a parameter-sensitive method for adding something to
    the build schedule.

    It takes as argument a file path, a directory path, a reference to a
    SCALAR containing perl code, or an IO::Handle (or subclass) object. It
    will retrieve code from the parameter as appropriate, parse it, and
    create zero or more Test::Inline::Script objects representing the test
    scripts that will be generated for that source code.

    Returns the number of test scripts added, which could be zero, or
    "undef" on error.

  add_class
      $Tests->add_class( 'Foo::Bar' );
      $Tests->add_class( 'Foo::Bar', recursive => 1 );

    The "add_class" method adds a class to the list of those to have their
    tests generated. Optionally, the "recursive" option can be provided to
    add not just the class you provide, but all classes below it as well.

    Returns the number of classes found with inline tests, and added,
    including 0 if no classes with tests are found. Returns "undef" if an
    error occurs while adding the class or it's children.

  add_all
    The "add_all" method will search the "InputHandler" for all *.pm files,
    and add them to the generation set.

    Returns the total number of test scripts added, which may be zero, or
    "undef" on error.

  classes
    The "classes" method returns a list of the names of all the classes that
    have been added to the "Test::Inline" object, or the null list "()" if
    nothing has been added.

  class
    For a given class name, fetches the Test::Inline::Script object for that
    class, if it has been added to the "Test::Inline" object. Returns
    "undef" if the class has not been added to the "Test::Inline" object.

  filenames
    For all of the classes added, the "filenames" method generates a map of
    the filenames that the test files for the various classes should be
    written to.

    Returns a reference to a hash with the classes as keys, and filenames as
    values.

    Returns 0 if there are no files to write.

    Returns "undef" on error.

  schedule
    While the "filenames" method generates a map of the files for the
    various classes, the "schedule" returns the list of file names in the
    order in which they should actually be executed.

    Returns a reference to an array containing the file names as strings.

    Returns 0 if there are no files to write.

    Returns "undef" on error.

  manifest
    The "manifest" generates the contents of the manifest file, if it is
    both wanted and needed.

    Returns the contents of the manifest file as a normal string, false if
    it is either not wanted or needed, or "undef" on error.

  save
      $Tests->save;

    The "save" method generates the test files for all classes, and saves
    them to the "output" directory.

    Returns the number of test files generated. Returns "undef" on error.

BUGS
    The "Extended =begin" syntax used for non-trivial sections is not really
    considered part of the spec yet. While simple '=begin testing' sections
    are find and will pass POD testing, extended begin sections may cause
    POD errors.

TO DO
    - Add support for "example" sections

    - Add support for "=for" sections

SUPPORT
    Bugs should always be submitted via the CPAN bug tracker

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Test-Inline>

    Professional support, assistance, or customisations for large scale uses
    of "Test::Inline" are available from <http://phase-n.com/>.

    For other issues, contact the maintainer.

AUTHOR
    Adam Kennedy <cpan@ali.as>, <http://ali.as/>

ACKNOWLEDGEMENTS
    Thank you to Phase N (<http://phase-n.com/>) for permitting the open
    sourcing and release of this distribution.

COPYRIGHT
    Copyright (c) 2004 - 2005 Phase N Austalia. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension Test::Inline

2.103    Wed Sep 28 2005
	- Removed Test::Inline::IO::File::VCS. It is incompleted, and never
	  should have been distributed in the dist.

2.102    Thu Sep 22 2005
	- To handle people with problems installing RT and a few other
	  programs, add a dependency on Pod::Tests

2.101    Tue Sep 13 2005
	- Rebuilding to synchronize with new Module::Install and
	  ExtUtils::AutoInstall
	- Tweaks the per-section test code to make more applicable value
	  appear in the failure summary.
	- Removed all remaining use of UNIVERSAL::isa, switching to a
	  more accurate Params::Util approach. Should allow for improved
	  extensibility.

2.100    Thu Jul 14 2005
	- Worked as good as I hoped, moving to production release
	- Bundling a couple of build-time-only deps

2.099_01 Tue Jul 12 2005
	- More work on making inline2test just DWYM
	- Added Test::Inline::Content::Simple

2.099_01 Sun Jul 10 2005
	- Starting to implement the 2.100 roadmap
	- Moved Test::Inline::Handler::Extract   to Test::Inline::Extract
	- Moved Test::Inline::Handler::File      to Test::Inline::IO::File
	- Moved Test::Inline::Handler::File::VCS to Test::Inline::IO::File::VCS
	- Convert inline2test to use a config file
	- Added Test::Inline::Content
	- Added Test::Inline::Content::Legacy
	- Added Test::Inline::Content::Default
	- Reorganised code to make it all work again
	- Added 12_content.t

2.002   Sat Jun 11 2005
	- The Test::Builder interface changes, breaking the test counts.
	- This change makes them work again

2.001   Sat Jun 11 2005
	- After determining Test::Inline 1 doesn't actually USE
	  the Test::Inline namespace, decided Schwern can just sod off
	  and I'm moving to production now.
	- Old users should not be effected if they specifically install
	  the module 'Pod::Test'

2.00_10 Sat Jun 11 2005
	- Making sure =begin testing SETUP behaves completely like the
	  previous =begin test setup sections

2.00_09 Fri Jun 10 2005
	- Adding Test::Inline::Handler::File::VCS to support automatically
	  adding output scripts to version control system, if they have one.
	- Added a --vcs flag to inline2test to use Test::Inline::Handler::File::VCS
	- Converted testing from =begin test setup to =begin testing SETUP

2.00_08 Fri May 21 2005
	- Adding a no_index to prevent bad things being indexed

2.00_07 Fri May 21 2005
	- Set inline2test to actually install
	- Remove pod2test, since I'm leaving it as is.

2.00_06 Thu May 20 2005
	- Added a few more bug fixes, and made the warning only a print,
	  and not a "true" warning.

2.00_05 Sat Mar 26 2005
	- Added support for =for example begin -> =for example end blocks

2.00_04	Fri Mar 25 2005
	- Resolved rt.cpan.org #557 (Catch and thow error for badly nested POD)

2.00_03 Mon Mar 21 2005
	- Forgot to update Changes in 2.00_02
	- Added automatic inclusion of 99_pod.t to my build script,
	  as recommended by CPANTS
	- Fixed a POD bug in Inline.pm

2.00_02 Mon Mar 21 2005
	- Moving some things around to make Test::Inline less
	  oriented around classes, and made adding things a bit
	  smarter.

2.00_01 Fri Mar 18 2005
	- Moved from PCPP (Phase N private CPAN repository) to Open Source
	  repository.
	- Inlined File::DirUtils as Test::Inline::Util
	- Renamed Test::Inline::TestFile as Test::Inline::Script
	- Completed POD docs for what methods exist
	- Partially completed main module DESCRIPTION

---------------------------------------------------------------------
Formerly known as Test::Inline2, a private Phase N Australia module

2.12 Tue Mar 15 2005
	- Fixed a bug in the generation of the manifest resulting from
	  other fixes in 2.11

2.11 Tue Mar 15 2005
	- Added 06_multifile to test a basic multi-file generation
	- Fixed so the we do actuall prepend the output path

2.10 Thu Mar 10 2005
	- Added support for the check_count flag, which allows the number
	  of tests run to be checked section-by-section, rather than
	  file-by-file only.

2.01 Mon Mar  7 2005
	- Tweaked some error messages

2.00 Thu Nov 18 2004
	- Mixed one final bug that was causing the manifest not to come out
	  in dependency order.

1.99 Thu Oct 28 2004
	- Doing error messages a slightly different way

1.98 Thu Oct 28 2004
	- Adding yet more error messages

1.97 Thu Oct 28 2004
	- Added more error messages (when in verbose mode)

1.96 Sun Sep  5 2004
	- Cleaned up Makefile.PL and dependencies

1.95 Sat Sep  4 2004
	- Fix in add_class

1.94 Fri Sep  3 2004
	- Fix a bug when generating manifests with relative paths

1.93 Fri Sep  3 2004
	- Save in sorted order for consistency

1.92 Fri Sep  3 2004
	- Fixed the bool and "" overloads for ::TestFile
	- Fixed manifest generation
	- Changed the meaning of the 'output' option slightly to make it more useful

1.91 Thu Sep  2 2004
	- Added the ability to supply a custom file content generator

1.90 Thu Sep  2 2004
	- More bugs that slipped through in ->save

1.64 Thu Sep  3 2004
	- Fixed a number of calls to ->_write
	- ->add and ->add_class both return the number of files added now

1.63 Thu Sep  2 2004
	- Made a small correction to the Test::Inline2::Handler::File API

1.62 Wed Sep  1 2004
	- Added a special hack to prevent one particular warning during a
	 unit test script in which we don't want to see it.

1.61 Web Sep  1 2004
	- Fixed some bugs in Test::Inline2::Handler::File

1.60 Wed Sep  1 2004
	- Changed to using Handler classes for the features we want to overload

1.50 Tue Aug 31 2004
	- Converted from Test::Inline::Heavy to Test::Inline2
	- Removed Test::Inline2::Config
	- Added class-level dependencies to a test section

1.02 Mon Aug 30 2004
	- Removed an erroneous $DB::single = 1

1.01 Fri Aug 20 2004
	- Updated Algorithm::Dependency to 1.02

1.00 Fri Aug 20 2004
	- Updating Algorithm::Dependency to 1.01 to fix critical ignore_orphans failure
	- Added a warning when there is a missing dependency

0.9  missing

0.8  missing

0.7  
	- Using File::Slurp to read files

0.6  missing

0.5  missing

0.4  Sat Apr 17 2004

lib/Test/Inline/Util.pm
# This is an inlined version of the private Phase N module File::DirUtils,
# approved for use only in its inline state as Test::Inline::Util.
# It will be released to CPAN at some later time, once complete.
# We ask that until that time you respect our development process and
# do not use this code.
package Test::Inline::Util;
use strict;
use File::Spec::Functions ':ALL';
use vars qw{$VERSION};
BEGIN {
$VERSION = '2.102';
}
sub shorten {
my $class = ref $_[0] ? ref shift : shift;
my $path = (defined $_[0] and length $_[0])
? canonpath( shift )
: return shift;
my @parts = splitdir( $path );
my $i = 0;
while ( defined $parts[++$i] ) {
next unless $i;
next unless $parts[$i] eq updir();
next if $parts[$i - 1] eq updir();
splice @parts, $i - 1, 2;
$i -= 2;
}
catdir( @parts );
}
sub parts {
my $class = ref $_[0] ? ref shift : shift;
my $path = $class->shorten(shift);
$path = '' if $path eq curdir();
scalar splitdir($path);
}
sub inverse {
my $class = ref $_[0] ? ref shift : shift;
my $path = $class->shorten( shift );
if ( ! defined $path or $path eq '' or $path eq curdir() ) {
return $path;
}
return undef if file_name_is_absolute( $path );
my @parts = splitdir( $path );
return undef if $parts[0] eq updir();
catdir( (updir()) x scalar @parts );
}
sub commonise {
my $class = ref $_[0] ? ref shift : shift;
my $first = $class->shorten( shift );
my $second = $class->shorten( shift );
return undef unless defined $first;
return undef unless defined $second;
my @first = splitdir( $first );
my @second = splitdir( $second );
my @base = ();
while ( defined $first[0] and defined $second[0] and $first[0] eq $second[0] ) {
push @base, $first[0];
shift @first;
shift @second;
}
[ catdir(@base), catdir(@first), catdir(@second) ];
}
sub relative {
my $class = ref $_[0] ? ref shift : shift;
my $commonised = $class->commonise( @_ ) or return undef;
my $from = $class->inverse( $commonised->[1] );
return undef unless defined $from;
my $to = $commonised->[2];
if ( $from eq '' and $to eq '' ) {
return '';
}
catdir( grep { length $_ } ($from, $to) );
}
1;


Makefile.PL
use inc::Module::Install;

# Generate configuration
name           ( 'Test-Inline'                        );
abstract       ( 'Inlining your tests next to the code being tested' );
author         ( 'Adam Kennedy <adam@ali.as>'         );
version_from   ( 'lib/Test/Inline.pm'                 );
requires       ( 'perl'                  => '5.005'   );

# Other people's CPAN modules
requires       ( 'File::Spec'            => '0.80'    );
requires       ( 'List::Util'            => '1.11'    );
requires       ( 'Getopt::Long'          => '2.34'    );
requires       ( 'File::Slurp'           => '9999.04' );
requires       ( 'File::Find::Rule'      => '0.26'    );

# My CPAN modules
requires       ( 'Config::Tiny'          => '2.00'    );
requires       ( 'Params::Util'          => '0.05'    );
requires       ( 'Class::Autouse'        => '1.15'    );
requires       ( 'Algorithm::Dependency' => '1.02'    );
requires       ( 'File::Flat'            => '0.95'    );

# Because some programs incorrectly do things like...
#   if ( installed('Test::Inline') ) { system('pod2test') }
# ... we need to make sure it is installed.
# This is a temporary arrangement to end in 2008.
requires       ( 'Pod::Tests'            => '0.18'    );

# Testing modules (bundled where appropriate)
build_requires ( 'Test::More'            => '0.47'    );
build_requires ( 'Test::ClassAPI'        => '1.02'    );
include        ( 'Test::ClassAPI'                     );
include        ( 'Class::Inspector'                   );

# Miscellaneous items
install_script ( 'bin/inline2test'                    );
no_index       ( directory               => 't.data'  );
auto_install   (                                      );

&WriteAll;
